### ISSUE TITLE
Inconsistent Equivalence Result for DCX Validation

### ISSUE THREAD

**Issue by MattePalte** [cda-tum/mqt-qcec#489]
### Environment information

- OS: Ubuntu 20.04  
- C++ Compiler: g++ (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0  
- mqt.core version: 2.7.0  
- mqt.qcec version: 2.7.1  
- Qiskit version: 1.2.4

### Description

The QCEC tool exhibits non-deterministic behavior when comparing the correct definition of the DCX gate with its incorrect decomposition. Approximately 90% of the time, QCEC incorrectly reports that the two circuits are equivalent, while 10% of the time it reports that it cannot draw any conclusion.

### Expected behavior

The QCEC tool should consistently report that the two circuits are not equivalent, as the incorrect decomposition of the DCX gate produces gives a circuit that is not semantically equivalent.

### How to Reproduce

1. Create two QASM files: one with the correct definition of the DCX gate and another with the incorrect decomposition.
2. Use the QCEC tool to compare the two circuits.
3. Run the comparison 1000 times and record the results.

**Minimal Example Code**
```python
from qiskit.qasm2 import loads
from qiskit import qasm2
from mqt import qcec
import tempfile
from qiskit.quantum_info import Statevector


def check_two_qasm_strings(qasm_str1, qasm_str2):
    # create two temporary files
    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix=".qasm") as f1, tempfile.NamedTemporaryFile(mode='w', delete=False, suffix=".qasm") as f2:
        f1.write(qasm_str1)
        f2.write(qasm_str2)
        f1.flush()
        f2.flush()
        # verify the two files
        res = qcec.verify(f1.name, f2.name)
        equivalence = str(res.equivalence)
        # print(f"Equivalence: {equivalence}")
        return equivalence


def remove_measurements_and_execute(qasm_str1, qasm_str2):
    # Load the circuits from the QASM strings
    circuit1 = loads(
        qasm_str1, custom_instructions=qasm2.LEGACY_CUSTOM_INSTRUCTIONS)
    circuit2 = loads(
        qasm_str2, custom_instructions=qasm2.LEGACY_CUSTOM_INSTRUCTIONS)

    # Remove measurements from the circuits
    circuit1.remove_final_measurements(inplace=True)
    circuit2.remove_final_measurements(inplace=True)

    # Execute them with statevector simulator
    sv_1 = Statevector.from_instruction(circuit1)
    probabilities_1 = sv_1.probabilities_dict()
    print(qasm_str1)
    print(circuit1)
    print("Statevector simulator result: ", probabilities_1)
    sv_2 = Statevector.from_instruction(circuit2)
    probabilities_2 = sv_2.probabilities_dict()
    print(qasm_str2)
    print(circuit2)
    print("Statevector simulator result: ", probabilities_2)


# minimal example
dcx_str = """
OPENQASM 2.0;
include "qelib1.inc";
gate dcx q0,q1 {
    cx q0,q1;
    cx q1,q0; }
qreg qr[2];
creg cr[2];
x qr[0];
dcx qr[0],qr[1];
measure qr[0] -> cr[0];
measure qr[1] -> cr[1];
"""

decomposed_str = """
OPENQASM 2.0;
include "qelib1.inc";
qreg qr[2];
creg cr[2];
x qr[0];
cx qr[1],qr[0];
cx qr[0],qr[1];
measure qr[0] -> cr[0];
measure qr[1] -> cr[1];
"""

answers = {}
for _ in range(1000):
    res = check_two_qasm_strings(
        dcx_str,
        decomposed_str
    )
    if res in answers.keys():
        answers[res] += 1
    else:
        answers[res] = 1

remove_measurements_and_execute(
    dcx_str,
    decomposed_str
)

answers
```
Output:
```plaintext
OPENQASM 2.0;
include "qelib1.inc";
gate dcx q0,q1 {
    cx q0,q1;
    cx q1,q0; }
qreg qr[2];
creg cr[2];
x qr[0];
dcx qr[0],qr[1];
measure qr[0] -> cr[0];
measure qr[1] -> cr[1];

      ┌───┐┌──────┐
qr_0: ┤ X ├┤0     ├
      └───┘│  Dcx │
qr_1: ─────┤1     ├
           └──────┘
Statevector simulator result:  {'10': 1.0}

OPENQASM 2.0;
include "qelib1.inc";
qreg qr[2];
creg cr[2];
x qr[0];
cx qr[1],qr[0];
cx qr[0],qr[1];
measure qr[0] -> cr[0];
measure qr[1] -> cr[1];

      ┌───┐┌───┐
qr_0: ┤ X ├┤ X ├──■──
      └───┘└─┬─┘┌─┴─┐
qr_1: ───────■──┤ X ├
                └───┘
Statevector simulator result:  {'11': 1.0}
{'equivalent': 960, 'no_information': 40}
```

**Expected Output**
The output of the comparison should consistently report that the two circuits are not equivalent.

**Actual Output**
The actual output shows that the QCEC tool reports that the two circuits are equivalent approximately 90% of the time and cannot draw any conclusion approximately 10% of the time. The output also shows the results of the comparison:
```python
{'equivalent': 960, 'no_information': 40}
```

**Possible Problem**
There could be an incorrect decomposition of the dcx gate in the QCEC parsing of the first QASM, or a wrong definition of the same dcx gate in the simulator used under the hood. 

Let me know if you need any additional information, I wish you a happy and productive day

**Comment by burgholzer** [comment]
Thanks for raising this! Definitely a bug.
Just very quickly looked at it and the reason for the (arguably strange) behaviour you are seeing is that:
The gate matrix definition that all of our decision diagram based checkers are using (defined here in MQT Core https://github.com/cda-tum/mqt-core/blob/cda641c03ed4aa018a49e0a2f703add831259006/include/mqt-core/dd/GateMatrixDefinitions.hpp#L93-L94) is based on a different qubit ordering convention. Thus all DD-based backend will yield `equivalent` although your circuits should not be equivalent.
The reason for getting `no_information` is quite interesting actually. This happens because we are running multiple different checks in parallel and it happens whenever the ZX-based equivalence checker and any of the DD-based checker finish at roughly the same time so that the other check is not aborted before finishing. 
The ZX-checker, which does not rely on the above matrix definition, but its own definition of how to translate a DCX gate to a ZX-diagram, gets it right. 
Eventually both checkers do not agree on the equivalence of the circuit and, thus, return "no_information".

This should be an easy fix in mqt-core, which I will probably get to within the next couple of days!
Thanks again for bringing this up!

**Comment by burgholzer** [comment]
Just released v2.8.1 with the fix: https://github.com/cda-tum/mqt-qcec/releases/tag/v2.8.1
Should be up on PyPI in a couple of hours.

Thanks again for reporting this 🙏🏼
